File: agentmail-sales-signal-router/typescript/src/watchlist.ts
/**
 * Watchlist loader — re-reads watchlist.json on every call so live edits
 * take effect without restarting the agent.
 */

import { readFileSync, existsSync } from "node:fs";

export interface Watchlist {
  deal_owners: Record<string, string>;
  watchlist_domains: string[];
  watchlist_keywords: string[];
  crm_notification_senders: string[];
}

const FILE = "watchlist.json";
const EMPTY: Watchlist = {
  deal_owners: {},
  watchlist_domains: [],
  watchlist_keywords: [],
  crm_notification_senders: [],
};

export function load(): Watchlist {
  if (!existsSync(FILE)) return EMPTY;
  try {
    const parsed = JSON.parse(readFileSync(FILE, "utf-8"));
    return {
      deal_owners: parsed.deal_owners ?? {},
      watchlist_domains: parsed.watchlist_domains ?? [],
      watchlist_keywords: parsed.watchlist_keywords ?? [],
      crm_notification_senders: parsed.crm_notification_senders ?? [],
    };
  } catch (e: any) {
    console.warn(`  ! watchlist.json malformed: ${e.message}`);
    return EMPTY;
  }
}

export function contextBlock(wl: Watchlist): string {
  const fmt = (xs: string[]) => xs.length ? xs.join(", ") : "(none)";
  return (
    `Watchlist domains: ${fmt(wl.watchlist_domains)}\n` +
    `Watchlist keywords: ${fmt(wl.watchlist_keywords)}\n` +
    `Known CRM/billing notification senders: ${fmt(wl.crm_notification_senders)}`
  );
}

export function findOwner(wl: Watchlist, senderEmail: string, body: string): string {
  const owners = wl.deal_owners ?? {};
  const domain = senderEmail.includes("@") ? senderEmail.split("@", 2)[1].toLowerCase() : "";
  if (domain && owners[domain]) return owners[domain];

  const bodyLower = body.toLowerCase();
  for (const [d, owner] of Object.entries(owners)) {
    if (d === "_comment") continue;
    if (bodyLower.includes(d.toLowerCase())) return owner;
  }
  return "";
}

File: agentmail-sales-signal-router/python/watchlist.py
"""
Watchlist loader — re-reads watchlist.json on every call so live edits take
effect without restarting the agent.
"""

import json
from pathlib import Path

WATCHLIST_FILE = Path("watchlist.json")


def load() -> dict:
    if not WATCHLIST_FILE.exists():
        return {
            "deal_owners": {},
            "watchlist_domains": [],
            "watchlist_keywords": [],
            "crm_notification_senders": [],
        }
    try:
        return json.loads(WATCHLIST_FILE.read_text())
    except Exception as e:
        print(f"  ! watchlist.json malformed: {e}")
        return {
            "deal_owners": {},
            "watchlist_domains": [],
            "watchlist_keywords": [],
            "crm_notification_senders": [],
        }


def context_block(wl: dict) -> str:
    """Render the watchlist as a readable block to inject into the user message."""
    domains = wl.get("watchlist_domains", []) or []
    keywords = wl.get("watchlist_keywords", []) or []
    senders = wl.get("crm_notification_senders", []) or []
    return (
        f"Watchlist domains: {', '.join(domains) if domains else '(none)'}\n"
        f"Watchlist keywords: {', '.join(keywords) if keywords else '(none)'}\n"
        f"Known CRM/billing notification senders: {', '.join(senders) if senders else '(none)'}"
    )


def find_owner(wl: dict, sender_email: str, body_text: str) -> str:
    """Return Slack member ID for the rep who owns this domain, or empty string."""
    owners = wl.get("deal_owners", {}) or {}
    domain = sender_email.split("@", 1)[-1].lower() if "@" in sender_email else ""
    if domain and domain in owners:
        return owners[domain]
    # Also check if any owned domain is mentioned in the body (e.g. forwarded thread)
    body_lower = body_text.lower()
    for d, owner in owners.items():
        if d == "_comment":
            continue
        if d.lower() in body_lower:
            return owner
    return ""
